Website Query API: Add `import-site` and `import-content` options

eliot-akira opened the pull request:
## What is this PR doing?

Adds query parameter `import-site` and `import-content` to the Playground website's Query API.

- `import-site` - Imports site files and database from a **zip file** specified by URL.
- `import-content` - Imports site content from a **WXR or WXZ file** specified by URL.

## What problem is it solving?

Resolves #608.

- #608

## Testing Instructions

- Start local server with `npm run dev`
- Visit the following URL and confirm **site files and database** have been imported:

  http://localhost:5400/website-server/?php=8.0&wp=6.4&storage=none&import-site=https://raw.githubusercontent.com/dmsnell/playground-experiments/trunk/just-a-little-change/playground.zip

- Visit the following URL and confirm **site content** has been imported:

  http://localhost:5400/website-server/?php=8.0&wp=6.4&storage=none&networking=yes&import-content=https://raw.githubusercontent.com/WordPress/theme-test-data/master/themeunittestdata.wordpress.xml

dmsnell commented on packages/playground/website/src/lib/make-blueprint.tsx:
```diff
@@ -17,11 +17,16 @@ export function makeBlueprint(options: MakeBlueprintOptions): Blueprint {
+			options.import
+				? {
+						step: 'importFile',
+						file: options.import,
```
do we want to make sure this only includes `http:` or `https:` schemas? I'm not sure if the browser would end up trying to download a `file://` link. hopefully not, but I don't want to open up the possibility of having a playground script lift a file from peoples' systems and then exporting it across the net.

eliot-akira replied:
Good idea, it makes sense to restrict the kind of value that the GET parameter passes. I'll add a step that ensures the protocol `http` and `https`.

That reminds me, if I understand correctly, this zip file must be served with CORS enabled - similar to plugins and themes, which require a proxy to get around CORS. In which case, it would be a good to document this restriction, since I imagine people will run into the issue, for example, linking to a zip package from a WordPress site's upload folder.

eliot-akira replied:
Done: [Ensure protocol http or https for import file URL](https://github.com/WordPress/wordpress-playground/pull/610/commits/a999e378276a9a3d565679f78f12d2e97f6c713e).

dmsnell commented on packages/playground/website/src/lib/make-blueprint.tsx:
```diff
@@ -17,11 +17,16 @@ export function makeBlueprint(options: MakeBlueprintOptions): Blueprint {
-				username: 'admin',
-				password: 'password',
-			},
+			options.import
```
I'm assuming this is not a problem for you, that you will address the type to add the import flag, but if not, I can try and add it in if you'd like.

eliot-akira replied:
Yes, I updated the interface `MakeBlueprintOptions` to include the new `import` option.

dmsnell reviewed (commented):
This is looking clean and small 👍

dmsnell commented on packages/playground/website/src/lib/make-blueprint.tsx:
```diff
@@ -17,11 +18,19 @@ export function makeBlueprint(options: MakeBlueprintOptions): Blueprint {
+						step: 'login',
+						username: 'admin',
+						password: 'password',
+				  },
```
while it makes sense that an import may have a different admin user, when we make this change it won't be easy for someone to auto-login to their imported site unless they add custom PHP to do so. I guess if they add an import they will need to manually add the login step (if they want it) and we should communicate that?

eliot-akira replied:
> an import may have a different admin user

Exactly, that was the thinking behind why the initial step is **either a site import *or* an auto-login step**, since we don't know that the admin user credentials will be the same as the default empty site.

> when we make this change it won't be easy for someone to auto-login to their imported site unless they add custom PHP to do so.

True, and crucially, there's currently no way to auto-login to an imported site from the Query API (which only has `login=1` or `0`).

A simplest naive solution might be, if we could assume that the admin user in the imported site has the same credentials as the default empty site. Then we can run **both import site *and* auto-login steps** initially on load. If the admin user is different, then auto-login can only happen through a custom Blueprint step and not through the Query API (currently).

If we had the concept of an optional ("fail-able") step, we could import a site then attempt to log in as default admin user, continuing to remaining steps even if the login fails.

dmsnell replied:
> no way to auto-login to an imported site from the Query API

I guess we do have this: passing blueprints as a URL fragment, `#{"steps":[{"step":"login","username":"admin","password":"password"}]}`

so maybe it's not as bad as I first thought. again, a bit verbose, but something someone can already do and we can make a builder UI to generate an appropriate URL.

> If we had the concept of an optional ("fail-able") step, we could import a site then attempt to log in as default admin user, continuing to remaining steps even if the login fails.

this sounds great! and a good use of the optional steps

adamziel replied:
@eliot-akira when the `login` step fails, it fails silently. That's not ideal and I'd love to introduce the concept of optional steps, but we could actually always try to log the user in as `admin`/`password` until the optional steps work lands. Would you please update this PR? I'd love to get this in.

dmsnell reviewed (commented):
can we also add an item to the table of options in the [Query API docs](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/docs/site/docs/08-query-api/01-index.md)?

dmsnell commented on packages/docs/site/docs/08-query-api/01-index.md:
```diff
@@ -27,6 +27,7 @@ You can go ahead and try it out. The Playground will automatically install the t
 | `wp`           | `latest`      | Loads the specified WordPress version. Supported values: `5.9`, `6.0`, `6.1`, `6.2`, `latest`                                                                                                                                                                                                                                                                  |
 | `plugin`       |               | Installs the specified plugin. Use the plugin name from the plugins directory URL, e.g. for a URL like `https://wordpress.org/plugins/wp-lazy-loading/`, the plugin name would be `wp-lazy-loading`. You can pre-install multiple plugins by saying `plugin=coblocks&plugin=wp-lazy-loading&…`. Installing a plugin automatically logs the user in as an admin |
 | `theme`        |               | Installs the specified theme. Use the theme name from the themes directory URL, e.g. for a URL like `https://wordpress.org/themes/disco/`, the theme name would be `disco`. Installing a theme automatically logs the user in as an admin                                                                                                                      |
+| `import`       |               | Imports a site from the specified zip file URL. The file must be served with CORS enabled. The step automatically logs the user in as an admin, unless the default user has been changed or option `login` is set to `0` (zero)                                                                                                                                |
```
> the file must be served with appropriate CORS headers (must allow fetching from the domain on which the Playground is running).

we might want to expand this, because "CORS enabled" may not say much. I _think_ the most important part is that a site allows the Playground domain as an allowable origin if using `Access-Control-Allow-Origin`.

dmsnell replied:
> The step automatically logs the user in as an admin, unless the default user has been changed or option `login` is set to `0` (zero)

this is great, and I'm glad you put it in, but don't we need to update our code to do this too?

if we don't want to modify the code at this time given the question it raises "how do we know if the login will succeed before trying?" then we could also note here something like "When importing a site, no auto-login will be performed because there's no guarantee that the admin user is the same as the default one set in the Playground. To auto-login a user add a `login` step in blueprints (see [[Blueprints in the hash]])"

dmsnell reviewed (commented):
it seems fine to allow `import` without auto-login, and we can add something better afterwards. your call is good to auto-login _or_ import given our options.

adamziel commented:
I left one minor comment @eliot-akira. The only other thing I'd change is I'd call this option `importFile` to be consistent with the related Blueprint step AND avoid confusion with the future `importWpContentDirectory` option (name TBD) I'm planning to add. Other than that, it looks great and I'd love to ship this, thank you! ❤️ It would help with https://github.com/WordPress/wordpress-playground/issues/718, too!

eliot-akira commented:
> The only other thing I'd change is I'd call this option `importFile`

OK, will do.

>  when the login step fails, it fails silently. That's not ideal..but we could actually **always try to log the user in** ..until the optional steps work lands. Would you please update this PR? 

Sure, I will make the change. Just to confirm my understanding: so for now it will import the site *and* attempt to auto-login with the default user credentials. Even if the login fails, the rest of the steps (like install theme and plugins) should continue to run.

When the "optional steps" feature is ready, then this login step can be made optional so that it doesn't throw an error  and stop the install (in case the imported site doesn't have the default admin user).

eliot-akira commented:
OK, I made the suggested changes.

- [Login with default admin user after importing site](https://github.com/WordPress/wordpress-playground/pull/610/commits/2f3e7d6ebcde68641127b462c76e2c418318a531)
- [Rename query parameter import to importFile](https://github.com/WordPress/wordpress-playground/pull/610/commits/f822a73b118248cb71e0efda1e2669919ef26614)

I noticed that the query parameter `login` was documented but not implemented (to disable auto-login with value `0`), so I added it - but it conflicted with the `trunk` branch which had it already. So I reverted that commit.

More importantly.. Recently I was catching up on new developments in the Playground repo, and learned that the `importFile` step was unintentionally removed in PR #780 (and reported in issue #833).

That means **this current PR shouldn't be merged until the `importFile` step is restored**.

But on that topic: `importFile` was the only step that is not "isomorphic" because it used `DOMParser`, which is not implemented in Node.js, to interact with the admin screen for WordPress Importer. Ideally, rather than restoring the step as it was, it might be better to write a new function that directly imports such files using `PHP.run()`. Maybe it could be temporarily restored until the new one is ready.

This also made me wonder, the name `importFile` is a bit ambiguous what kind of file it accepts. Looking at the last existing state of that step [here](https://github.com/WordPress/wordpress-playground/blob/00305f46d10f793e719bfb7a42eb00373ca82b50/packages/playground/blueprints/src/lib/steps/import-export.ts#L184-L192), I see it expects a WXR/WXZ file. Previously I thought it accepted a zip export of the whole site - hence the title of this PR, "install site from zip file URL". The original issue that prompted this PR also meant this:

- #608

In this case I suppose the query parameter would be better named `importSite`. (Or better yet, import only the contents of `wp-content` as in the new `importWordPressFiles` step.)

adamziel commented:
> That means this current PR shouldn't be merged until the importFile step is restored.

Oh you're right! I just started a PR to restore the importFile step https://github.com/WordPress/wordpress-playground/pull/835/

> But on that topic: importFile was the only step that is not "isomorphic" because it used DOMParser, which is not implemented in Node.js

I wonder what would it take to call the appropriate PHP functions directly, I like that idea. Alternatively, we could use the new `WP_HTML_Tag_Processor` on PHP side to replace the DOMParser. 

> In this case I suppose the query parameter would be better named importSite. (Or better yet, import only the contents of wp-content as in the new importWordPressFiles step.)

I like the idea about reorganizing the names of these steps. `importSite` is also a bit ambiguous but it's better than `importFile`. How about splitting the `importFile` step into two distinct steps called `importWxr`, `importWxz`, and then renaming `importWpFiles` to `importWpContentDirectory`?

eliot-akira commented:
> I wonder what would it take to call the appropriate PHP functions directly, I like that idea. Alternatively, we could use the new WP_HTML_Tag_Processor on PHP side to replace the DOMParser.

It probably depends on how [WordPress Importer](https://github.com/WordPress/wordpress-importer) works internally, if its import function is written in a way that's callable independently and programmatically, or if it's tightly coupled with the admin screen and form submit flow. In the latter case, maybe the majority of the existing `importFile` step can be kept while replacing DOMParser with HTML tag processor, which is a great idea.

> How about splitting the `importFile` step into two distinct steps called `importWxr`, `importWxz`

That sounds good to me, to include the expected file format in the step name. Or maybe `importContent` step that accepts both WXR/WXZ formats, distinguished by the file extension.

> and then renaming `importWpFiles` to `importWpContentDirectory`?

For this one, if there's a single common format to export/import a site (in its entirety or just the `wp-content` directory), then I kinda still like `importSite` as a short way to describe what it does. By chance I found the original pull request where the import/export functions were implemented.

- https://github.com/WordPress/wordpress-playground/pull/209

So there are ways to:

- Export and import the whole site (all files and database): `zipEntireSite` (previously called `exportFile`) and `replaceSite`
- Export and import the site content (partial database): `exportWXR`, `exportWXZ`, and `submitImporterForm`

Looking at this, I can imagine there may be other export formats in the future, like: a "site diff" for only the changed files and database operations; or some kind of JSON format to create content; or importing an SQL file. In that case, it may be better to have step/function names that specify the expected file format, instead of abstract terms like "content" and "site".

I tend to err on the side of user convenience, short and easy to type, like an `import` step that accepts any supported export format (WXR/WXZ/ZIP/JSON/SQL..) - as they say, "be liberal in what you accept", the so-called Robustness Principle - but sometimes that can cause ambiguity and confusion, or restrict future extensibility, so I understand there's value in being explicit, specific, and even verbose.

aehlke commented:
For user research you may look at what specific import/export options popular cloud wordpress providers offer. I believing they typically have ways to replace sites (between staging/prod etc) as well as maybe ways to sync content across environments without necessarily syncing full site config but I'm rusty on WP now

adamziel commented:
@eliot-akira this is unblocked now as https://github.com/WordPress/wordpress-playground/pull/835 is merged.

> importContent step that accepts both WXR/WXZ formats, distinguished by the file extension.

`importContent` sounds great! Let's do that. It also handles both wxr and wxz using the same admin page so maybe it doesn't make as much sense to split that logic in two separate steps.

> I kinda still like importSite as a short way to describe what it does

That's fair, and I guess we'll need to go beyond `wp-content` anyway as, say, `wp-config.php` lives in WordPress root. Alright, let's settle for `importContent` and `importSite`.

> I tend to err on the side of user convenience, short and easy to type, like an import step that accepts any supported export format (WXR/WXZ/ZIP/JSON/SQL..) 

I like that idea! That could be an entirely new step that's a higher-level abstraction – it knows how to detect the input format, and then call a specific format-specific step.

eliot-akira commented:
> Alright, let's settle for `importContent` and `importSite`.
>
> [The `import` step] could be an entirely new step that's a higher-level abstraction – it knows how to detect the input format, and then call a specific format-specific step.

That sounds nice for a public interface like the Query API.

OK, I'll work on the above points. It sounds simple enough, as a thin layer to create a blueprint step from the given query parameter. (I guess these steps cannot be used together? Maybe `importSite` gets priority, and if there's `importContent` specified, it can run after it.)

---

> For user research you may look at what specific import/export options popular cloud wordpress providers offer.

@aehlke - Thank you for the suggestion. It's a good idea to see what existing import/export solutions there are.

It reminds me of WP-CLI, how it has commands to import/export the database, and particularly its [`search-replace` command](https://github.com/wp-cli/search-replace-command) which is necessary to update the site URL after importing a new database. Also, having such a blueprint step could help solve an issue like this one, which is caused by hardcoded site URL in the database:

- WordPress/playground-tools#14 

It also makes me wonder whether it's possible in the future to run blueprint steps on regular WP sites outside of Playground. As the Blueprint API evolves, the more it seems like a useful abstraction for setting up a WordPress site in any context, including `wp-env` and for general-purpose use.

adamziel commented:
> It sounds simple enough, as a thin layer to create a blueprint step from the given query parameter. (I guess these steps cannot be used together? Maybe `importSite` gets priority, and if there's `importContent` specified, it can run after it.)

Actually I’m having second thoughts here. It’s easy to pattern match the file extension, but then extending support to, say, plugins or different site export formats could require unzipping the archive and reasoning about its contents — which feels like a big scope creep.

eliot-akira commented:
> plugins or different site export formats could require unzipping the archive and reasoning about its contents

True, there could be multiple export formats that use the `zip` file extension. I agree, guessing the export format from zip contents doesn't seem like a good idea.

If we can't rely on file extension to tell us the expected format - that means, instead of `importSite` and `importContent`, we should let the user specify the format in the step name (and query parameter name). Maybe..

- importWxr
- importWxz
- importSiteZip
- importSiteDiffZip
- importWpContentZip
- importPluginZip
- importContentJson

It does feel a bit redundant to have the file extension in the step/query name, if they are the same.

How about, if in the future we have different site export formats, their corresponding import/export step names could specify what they are..?

adamziel commented:
> How about, if in the future we have different site export formats, their corresponding import/export step names could specify what they are..?

I like that!

eliot-akira commented:
OK, the following options have been added to the website's Query API.

- `import-site` - Imports site files and database from a zip file specified by URL.

  Example: http://localhost:5400/website-server/?php=8.0&wp=6.4&storage=none&import-site=https://raw.githubusercontent.com/dmsnell/playground-experiments/trunk/just-a-little-change/playground.zip

- `import-content` - Imports site content from a WXR or WXZ file specified by URL.
  
  Example: http://localhost:5400/website-server/?php=8.0&wp=6.4&storage=none&networking=yes&import-content=https://raw.githubusercontent.com/WordPress/theme-test-data/master/themeunittestdata.wordpress.xml

(These example URLs can be visited after running `npm run dev`.)

---

Also updated the documentation page to describe these options. Some points to note:

- `import-site` (zip) and `import-content` (wxr/wxz) options can be used together. The steps are run in this order:
  - first it imports the site,
  - then logs in with the default admin user
  - then imports content.
- `import-content` option uses the WordPress Importer, so the default admin user must be logged in. It means the `login` option must be `1` (default) and the login must succeed.
- Added note about CORS policy

![image](https://github.com/WordPress/wordpress-playground/assets/5352434/4f74073b-8275-40b6-9aae-2960a237769b)

From what I understand, this applies to all query parameters that accept a URL, namely: `blueprint-url`, `plugin`, `theme`, `import-site`, and `import-content`.

adamziel commented on packages/docs/site/docs/08-query-api/01-index.md:
```diff
@@ -33,11 +33,19 @@ You can go ahead and try it out. The Playground will automatically install the t
+
+:::info CORS policy
+
+To import files from a URL, such as a site zip package, they must be served with `Access-Control-Allow-Origin` header set. For reference, see: [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers).
```
Lovely note, thank you @eliot-akira! ❤️

adamziel commented on packages/docs/site/docs/08-query-api/01-index.md:
```diff
@@ -33,11 +33,19 @@ You can go ahead and try it out. The Playground will automatically install the t
 | `mode`                 | `seamless`, `browser`, or `browser-full-screen` | Displays WordPress on a full-page or wraps it in a browser UI                                                                                                                                                                                                                                                                      ...
 | `lazy`                 |                                                 | Defer loading the Playground assets until someone clicks on the "Run" button                                                                                                                                                                                                                                                       ...
-| `login`                | `1`                                             | Logs the user in as an admin                                                                                                                                                                                                                                                                                                       ...
+| `login`                | `1`                                             | Logs the user in as an admin. Set to `0` to not log in.                                                                                                                                                                                                                                                                            ...
```
Good catch, thank you @eliot-akira! I know it's not directly related to this PR, but it's such a small and useful change – let's keep it.

adamziel commented:
- [x] The testing instructions work
- [x] The change works
- [x] The tests are passing
- [x] The doc was updated

This is amazing @eliot-akira, thank you so much! ❤️
